pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentReplicator.java

```diff
@@ -134,30 +137,44 @@ public PersistentReplicator(String localCluster, PersistentTopic localTopic, Man
     }
 
     @Override
-    protected void readEntries(Producer<byte[]> producer) {
-        // Rewind the cursor to be sure to read again all non-acked messages sent while restarting
+    protected void setProducerAndTriggerReadEntries(Producer<byte[]> producer) {
+        // Rewind the cursor to be sure to read again all non-acked messages sent while restarting.
         cursor.rewind();
-
         cursor.cancelPendingReadRequest();
-        HAVE_PENDING_READ_UPDATER.set(this, FALSE);
-        this.producer = (ProducerImpl) producer;
```

Comment:
Why do we need to move these two lines?
It seems that if we don't set producer, the operation `doCloseProducerAsync(producer, () -> {}); ` can't close the producer.

Reply:
> Why do we need to move these two lines?
It seems that if we don't set producer, the operation doCloseProducerAsync(producer, () -> {});  can't close the producer.

We did not remove these two lines, just moved them into the logic branch `compareSetAndGetState(Starting, Started) == true`.

>`doCloseProducerAsync(producer, () -> {});`  can't close the producer.

This logic branch is using the method variable `producer`, not `this.producer`, so it does work

Reply:
OK